### PR TITLE
Specific BMH scale-down

### PR DESCRIPTION
**Pull request by abays**
Allows targeted BaremetalHost scale-down (via annotation) for BaremetalSets

### CODE REVIEW

File: controllers/baremetalset_controller.go [hunk 1]
```diff
@@ -274,32 +274,67 @@ func (r *BaremetalSetReconciler) ensureBaremetalHosts(instance *ospdirectorv1bet
...
 	// Find the current BaremetalHosts belonging to this BaremetalSet
 	for _, baremetalHost := range baremetalHostsList.Items {
 		if baremetalHost.Spec.ConsumerRef != nil && (baremetalHost.Spec.ConsumerRef.Kind == instance.Kind && baremetalHost.Spec.ConsumerRef.Name == instance.Name && baremetalHost.Spec.ConsumerRef.Namespace == instance.Namespace) {
 			r.Log.Info(fmt.Sprintf("Existing BaremetalHost: %s", baremetalHost.ObjectMeta.Name))
-			existingBaremetalHosts = append(existingBaremetalHosts, baremetalHost.ObjectMeta.Name)
+			// Storing the existing BaremetalHosts in a map like this just makes selected replica scale-down easier below
+			existingBaremetalHosts[baremetalHost.ObjectMeta.Name] = baremetalHost.ObjectMeta.Name
+
+			// Prep for possible replica scale-down below
+			if val, ok := baremetalHost.Annotations[baremetalset.BaremetalHostRemovalAnnotation]; ok && (val == "yes" || val == "true") {
```
Comment (stuggi) [inline controllers/baremetalset_controller.go:288]: we probably should use https://golang.org/pkg/strings/#EqualFold or strings.ToLower(val) to do case-insensitivity  compare, just to make sure someone did not set BaremetalHostRemovalAnnotation to Yes/True?

File: controllers/baremetalset_controller.go [hunk 2]
```diff
@@ -274,32 +274,67 @@ func (r *BaremetalSetReconciler) ensureBaremetalHosts(instance *ospdirectorv1bet
...
-				existingBaremetalHosts = []string{}
+				removalAnnotatedBaremetalHosts = []string{}
+			}
+
+			// We removed a removal-annotated BaremetalHost, so continue
+			continue
+		}
+
+		// We did NOT find a removal-annotated BaremetalHost to remove above, so we pick the first one
+		// in the existingBaremetalHosts list
```
Comment (stuggi) [inline controllers/baremetalset_controller.go:328]: We probably should not do any action if the host to remove was not annotated for removal to make sure we do not remove a node which is not expected to be deleted. The workflow today in tripleo also requires specify a specific node to delete.

File: controllers/baremetalset_controller.go (hunk 1 above)
Comment (abays) [reply controllers/baremetalset_controller.go:288]: Sure, that's a good point

File: controllers/baremetalset_controller.go [hunk 3]
```diff
@@ -274,33 +274,67 @@ func (r *BaremetalSetReconciler) ensureBaremetalHosts(instance *ospdirectorv1bet
...
-	// Deallocate existing BaremetalHosts to match the requested replica count, if necessary.
-	// TODO: Add something to CR to allow user to specify which BaremetalHosts are removed
-	for i := len(existingBaremetalHosts); i > instance.Spec.Replicas; i-- {
-		for _, oldBmhName := range existingBaremetalHosts {
-			// Just delete the first one we find until the TODO from above is implemented
-			err := r.baremetalHostDeprovision(instance, oldBmhName)
+	// Deallocate existing BaremetalHosts to match the requested replica count, if necessary.  First we
+	// choose BaremetalHosts with the "osp-director.openstack.org/baremetalset-delete-baremetalhost=yes"
+	// annotation, then, if there are still BaremetalHosts left to deprovision, we choose randomly from the
+	// existingBaremetalHosts map (as maps are unordered in Golang, we don't know which one will be chosen)
```
Comment (stuggi) [inline controllers/baremetalset_controller.go:297]: we should remove the choose randomly part of the comment.

**Review by stuggi — COMMENTED** [review]
lgtm, just one small change to remove the "random" part from the comment

File: controllers/baremetalset_controller.go (hunk 3 above)
Comment (abays) [reply controllers/baremetalset_controller.go:297]: Good catch

**Review by stuggi — APPROVED** [review]
lgtm

**Comment by dprince** [comment]
lgtm

**Pull request merged**